doc/source/reference/arrays.classes.rst

```diff
@@ -39,77 +39,119 @@ Special attributes and methods
+   which knows what to do with your class).
+
+   The presence of :func:`__array_ufunc__` also influences how
+   :class:`ndarray` handles binary operations like ``arr + obj`` and ``arr
+   < obj`` when ``arr`` is an :class:`ndarray` and ``obj`` is an instance
+   of a custom class. There are two possibilities. If
+   ``obj.__array_ufunc__`` is present and not :obj:`None`, then
+   ``ndarray.__add__`` and friends will delegate to the ufunc machinery,
+   meaning that ``arr + obj`` becomes ``np.add(arr, obj)``, and then
+   :func:`~numpy.add` invokes ``obj.__array_ufunc__``. This is useful if you
+   want to define an object that acts like an array.
+
+   Alternatively, if ``obj.__array_ufunc__`` is set to :obj:`None`, then as a
+   special case, special methods like ``ndarray.__add__`` will notice this
+   and *unconditionally* return :obj:`NotImplemented`, so that Python will
+   dispatch to ``obj.__radd__`` instead. This is useful if you want to define
+   a special object that interacts with arrays via binary operations, but
+   is not itself an array. For example, a units handling system might have
+   an object ``m`` representing the "meters" unit, and want to support the
+   syntax ``arr * m`` to represent that the array has units of "meters", but
+   not want to otherwise interact with arrays via ufuncs or otherwise. This
+   can be done by setting ``__array_ufunc__ = None`` and defining ``__mul__``
+   and ``__rmul__`` methods. (Note that this means that writing an
+   ``__array_ufunc__`` that always returns :obj:`NotImplemented` is not
+   quite the same as setting ``__array_ufunc__ = None``: in the former
+   case, ``arr + obj`` will raise :exc:`TypeError`, while in the latter
+   case it is possible to define a ``__radd__`` method to prevent this.)
+
+   The above does not hold for in-place operators, for which :class:`ndarray`
+   never returns :obj:`NotImplemented`.  Hence, ``arr += obj`` would always
+   lead to a :exc:`TypeError`.  This is because for arrays in-place operations
+   cannot generically be replaced by a simple reverse operation.  (For
+   instance, by default, ``arr[:] += obj`` would be translated to ``arr[:] =
+   arr[:] + obj``, which would likely be wrong.)
+
+   .. note:: If you define ``__array_ufunc__``:
+
+      - If you are not a subclass of :class:`ndarray`, we recommend your
+        class define special methods like ``__add__`` and ``__lt__`` that
+        delegate to ufuncs just like ndarray does.  An easy way to do this
+        is to subclass from :class:`~numpy.lib.mixins.NDArrayOperatorsMixin`.
+      - If you subclass :class:`ndarray`, we strongly recommend that you
+        avoid confusion by neither setting :func:`__array_ufunc__` to
+        :obj:`None`, which makes no sense for an array subclass, nor by
+        defining it and also defining reverse methods, which methods will
+        be called by ``CPython`` in preference over the :class:`ndarray`
+        forward methods.
```

Comment:
I tried to clarify:
```
      - If you subclass :class:`ndarray`, we recommend that you put all your
        override logic in ``__array_ufunc__`` and not also override special
        methods. This ensures the class hierarchy is determined in only one
        place rather than separately by the ufunc machinery and by ``CPython``
        (which gives preference to special methods of subclasses; the
        alternative way to enforce a one-place only, of setting
        :func:`__array_ufunc__` to :obj:`None`, would seem very unexpected and
        thus confusing, as then the subclass would not work at all with
        ufuncs).
```